internal/controllers/core/portforward/reconciler.go

```diff
@@ -128,23 +126,92 @@ func (r *Reconciler) startPortForwardLoop(ctx context.Context, entry portForward
 	}
 }
 
-func (r *Reconciler) onePortForward(ctx context.Context, entry portForwardEntry, forward Forward) error {
-	ns := k8s.Namespace(entry.Spec.Namespace)
-	podID := k8s.PodID(entry.Spec.PodName)
+func (r *Reconciler) updateForwardStatus(ctx context.Context, entry *portForwardEntry) {
+	var pf v1alpha1.PortForward
+	key := apis.Key(entry.PortForward)
+	if err := r.ctrlClient.Get(ctx, key, &pf); err != nil {
+		if !apierrors.IsNotFound(err) {
+			// short of dispatching a fatal error, there's nothing that can really be done here, so just log it
+			// for debugging purposes
+			logger.Get(ctx).Debugf("Failed to fetch PortForward %q for status update: %v", entry.Name, err)
+		}
+		return
+	}
+
+	newStatuses := entry.statuses()
+	if equality.Semantic.DeepEqual(pf.Status.ForwardStatuses, newStatuses) {
+		// the forwards didn't actually change, so skip the update
+		return
+	}
+
+	pf.Status.ForwardStatuses = newStatuses
+	if err := r.ctrlClient.Status().Update(ctx, &pf); err != nil {
+		if !apierrors.IsNotFound(err) && !apierrors.IsConflict(err) {
+			logger.Get(ctx).Debugf("Failed to update status for PortForward %q: %v", entry.Name, err)
+		}
+	}
+}
 
-	pf, err := r.kClient.CreatePortForwarder(ctx, ns, podID, int(forward.LocalPort), int(forward.ContainerPort), forward.Host)
+func (r *Reconciler) onePortForward(ctx context.Context, entry *portForwardEntry, forward Forward) {
+	pf, err := r.kClient.CreatePortForwarder(
+		ctx,
+		k8s.Namespace(entry.Spec.Namespace),
+		k8s.PodID(entry.Spec.PodName),
+		int(forward.LocalPort),
+		int(forward.ContainerPort),
+		forward.Host)
 	if err != nil {
-		return err
+		shouldUpdate := entry.setStatus(forward, ForwardStatus{
+			LocalPort:     forward.LocalPort,
+			ContainerPort: forward.ContainerPort,
+			Error:         err.Error(),
+		})
+		if shouldUpdate {
+			r.updateForwardStatus(ctx, entry)
+		}
+		return
 	}
 
+	// wait in the background for the port forwarder to signal that it's ready to update the status
+	// the doneCh ensures we don't leak the goroutine if ForwardPorts() errors out early without
+	// ever becoming ready
+	doneCh := make(chan struct{}, 1)
+	go func() {
+		readyCh := pf.ReadyCh()
+		if readyCh == nil {
+			return
+		}
+		select {
+		case <-doneCh:
```

Comment:
should we also select on ctx.Done()?

Reply:
_Shouldn't_ matter (if context is canceled, `ForwardPorts()` should return, so `doneCh` will get closed) but can't hurt anything and might avoid a nasty surprise in the future if we refactor - added in 83f8bc3a3f40d5a00d4eb3e3630a2d9ee58ca524